File: Javascript/Two Pointers/Valid-Palindrome.js
/**
 * @param {string} s
 * @return {boolean}
 */
const isPalindrome = (s) => {
  let left = 0;
  let right = s.length - 1;

  while (left < right) {
    let [leftChar, rightChar] = [s[left].toLowerCase(), s[right].toLowerCase()];
    if (isNonAlphaNumeric(leftChar)) {
      left++;
      continue;
    }
    if (isNonAlphaNumeric(rightChar)) {
      right--;
      continue;
    }

    if (leftChar != rightChar) {
      return false;
    }

    left++;
    right--;
  }

  return true;
};

const isNonAlphaNumeric = (char) => {
  const isNonAlpha = char < "a" || "z" < char; // a(97) - z(122)
  const isNonNumeric = char < "0" || "9" < char; // 0(48) - 9(57)

  return isNonAlpha && isNonNumeric;
};

File: Rust/Two Pointers/Valid-Palindrome.rs

impl Solution {
    pub fn is_palindrome(s: String) -> bool {
        let s: String = s
            .chars()
            .filter(|ch| ch.is_alphanumeric())
            .map(|ch| ch.to_ascii_lowercase())
            .collect();
        let half = s.len() / 2;
        let left = s[..half].chars();
        let right = s[(s.len() - half)..].chars();
        left.zip(right.rev()).all(|(a, b)| a == b)
    }
}

